### ISSUE TITLE
Error: VAE Tutorial in TF 2.2

### ISSUE THREAD

**Issue by anirban-mukherjee** [rstudio/keras#1058]
The RStudio Keras VAE tutorial (https://keras.rstudio.com/articles/examples/variational_autoencoder.html) bug checks at model training (i.e., immediately after calling keras::fit, right before the visualisation section of the tutorial).

Information about my system:
- Mac OS X 10.15.5, Python 3.8.2, TF 2.2, R 4.0.0. 
- R, Python, Reticulate, and Tensorflow are all working as expected. 
- Ran the code in the console in RStudio 1.3.1011. 
- I copy and pasted the tutorial code from the webpage into a new session.

Please let me know if you need any further information. Traceback below:

Error in py_call_impl(callable, dots$args, dots$keywords) : 
  _SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors, but found [<tf.Tensor 'dense_2/Identity:0' shape=(None, 2) dtype=float32>, <tf.Tensor 'dense_1/Identity:0' shape=(None, 2) dtype=float32>] 
22.
stop(structure(list(message = "_SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors, but found [<tf.Tensor 'dense_2/Identity:0' shape=(None, 2) dtype=float32>, <tf.Tensor 'dense_1/Identity:0' shape=(None, 2) dtype=float32>]", 
    call = py_call_impl(callable, dots$args, dots$keywords), 
    cppstack = structure(list(file = "", line = -1L, stack = c("1   reticulate.so                       0x0000000112165adf _ZN4Rcpp9exceptionC2EPKcb + 207", 
    "2   reticulate.so                       0x000000011216e1b5 _ZN4Rcpp4stopERKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE + 53",  ... 
21.
quick_execute at execute.py#72
20.
call at function.py#593
19.
_call_flat at function.py#1745
18.
_filtered_call at function.py#1661
17.
__call__ at function.py#2420
16.
_call at def_function.py#644
15.
__call__ at def_function.py#580
14.
fit at training.py#848
13.
_method_wrapper at training.py#66
12.
(structure(function (...) 
{
    dots <- py_resolve_dots(list(...))
    result <- py_call_impl(callable, dots$args, dots$keywords) ... 
11.
do.call(object$fit, args) 
10.
fit.keras.engine.training.Model(., x_train, x_train, shuffle = TRUE, 
    epochs = epochs, batch_size = batch_size, validation_data = list(x_test, 
        x_test)) 
9.
fit(., x_train, x_train, shuffle = TRUE, epochs = epochs, batch_size = batch_size, 
    validation_data = list(x_test, x_test)) 
8.
function_list[[k]](value) 
7.
withVisible(function_list[[k]](value)) 
6.
freduce(value, `_function_list`) 
5.
`_fseq`(`_lhs`) 
4.
eval(quote(`_fseq`(`_lhs`)), env, env) 
3.
eval(quote(`_fseq`(`_lhs`)), env, env) 
2.
withVisible(eval(quote(`_fseq`(`_lhs`)), env, env)) 
1.
vae %>% fit(x_train, x_train, shuffle = TRUE, epochs = epochs, 
    batch_size = batch_size, validation_data = list(x_test, x_test))

**Comment by henry090** [comment]
Hi. This could be due to tensorflow 2.2. I tested the code with TF 2.0/2.1 and it works fine.

**Comment by anirban-mukherjee** [comment]
If you comment out the line of code that computes the KL loss, the code doesn’t bug check and trains the network. I suspect there is some API change that is breaking one or more of the backend functions that are used in the line of code that computes the KL loss.

**Comment by anirban-mukherjee** [comment]
I believe this issue relates to https://github.com/tensorflow/tensorflow/issues/36398.

**Comment by anirban-mukherjee** [comment]
@skeydan : Do you think that your suggestion for https://github.com/rstudio/keras/issues/1008 may work in the VAE example? Perhaps, z_mean and z_log_var can be concatenated with x_decoded_mean and then all of that used to construct the loss?

**Comment by anirban-mukherjee** [comment]
@skeydan : Your idea worked like a charm. I modified the code so that the output of the Keras model was a list of three elements: z_mean, z_log_var, and x_decoded_mean. In the loss function, I used "[[" to access each of these three list elements and wrote a single line of code to compute the loss function. I think you're right that the key is that one cannot access the output of the "raw" layers of the network. Instead, one has to "pipe" the output of the layers out to the loss function as a model output and then write the loss function on the output. This is a cleaner and more explicit way to do things but it does require re-writing a little bit of code.

**Comment by skeydan** [comment]
Hi, 
glad to hear you got this working! I finally got to looking into this some more and found that the general workaround - 

```
experimental_run_tf_function=FALSE
```

in `compile` -

does not work anymore in TF 2.2. What *seems* to work is passing 
```
run_eagerly = TRUE
```

in `fit`, but this makes things extremely slow and certainly cannot be recommended.

Looking at https://github.com/tensorflow/tensorflow/issues/34944, it seems like this is not going to be fixed on the Python side, and that the recommendation in these cases is to rewrite the code to use `GradientTape`.

See also the comments on the top of this example, where we recommend to have a look at the more recent examples.

For anyone interested in following this recommendation, the best way as of today is to take a VAE example like https://github.com/rstudio/keras/blob/master/vignettes/examples/eager_cvae.R, which can again be optimized by using the fitting code from https://tensorflow.rstudio.com/tutorials/advanced/.

Here, for quick reference, is a code template for a generic autoencoder (an LSTM autoencoder but that's just because that's what I copied it from :-)), which you could adapt to your needs:

```
library(keras)
library(tfdatasets)
library(tfautograph)
library(reticulate)
library(purrr)

x_train <- ...

ds_train <- tensor_slices_dataset(x_train) %>%
  dataset_batch(batch_size)


encoder_model <- function(...,
                          name = NULL) {
  
  keras_model_custom(name = name, function(self) {
    
    self$noise <- ...
    self$lstm <- ...
    self$batchnorm <- ...
    
    function (x, mask = NULL) {
      x %>%
        self$noise() %>%
        self$lstm() %>%
        self$batchnorm() 
    }
  })
}

decoder_model <- function(...,
                          name = NULL) {
  
  keras_model_custom(name = name, function(self) {
    
    self$repeat_vector <- ...
    self$noise <- ...
    self$lstm <- ...
    self$batchnorm <- ...
    self$elu <- ... 
    self$time_distributed <- t...
    
    function (x, mask = NULL) {
      x %>%
        self$repeat_vector() %>%
        self$noise() %>%
        self$lstm() %>%
        self$batchnorm() %>%
        self$elu() %>%
        self$time_distributed()
    }
  })
}


encoder <- encoder_model(...)

decoder <- decoder_model(...)

optimizer <- optimizer_adam(lr = 1e-3)

mse_loss <- tf$keras$losses$MeanSquaredError(reduction = tf$keras$losses$Reduction$SUM)

train_loss <- tf$keras$metrics$Mean(name='train_loss')


train_step <- function(batch) {
  
  with (tf$GradientTape(persistent = TRUE) %as% tape, {
    code <- encoder(batch)
    reconstructed <- decoder(code)

    loss <- mse_loss(batch, reconstructed) 
  })
  
  encoder_gradients <- tape$gradient(loss, encoder$trainable_variables)
  decoder_gradients <- tape$gradient(loss, decoder$trainable_variables)
  
  optimizer$apply_gradients(
    purrr::transpose(list(encoder_gradients, encoder$trainable_variables))
  )
  optimizer$apply_gradients(
    purrr::transpose(list(decoder_gradients, decoder$trainable_variables))
  )
  
  train_loss(loss)
}

training_loop <- tf_function(autograph(function(ds_train) {
  
  for (batch in ds_train) {
    train_step(batch)
  }
  
  tf$print("Loss: ", train_loss$result())
  
  train_loss$reset_states()
  
}))

for (epoch in 1:200) {
  cat("Epoch: ", epoch, " -----------\n")
  training_loop(ds_train)  
}


```

HTH,
Sigrid
